common/nglinks.py

```diff
@@ -0,0 +1,116 @@
+"""Neuroglancer link functions."""
+import json
+import urllib
+from typing import Optional, Union
+from collections import OrderedDict
+from dataclasses import dataclass
+from enum import Enum
+
+
+class LayerType(Enum):
+    Image = "image"
+    Segmentation = "segmentation"
+
+
+@dataclass
+class Layer:
+    cloudpath: str
+    type: LayerType
+
+
+class ImageLayer(Layer):
+
+    def __init__(self, cloudpath: str):
+        self.cloudpath = cloudpath
+        self.type = LayerType.Image
+
+
+class SegLayer(Layer):
+
+    def __init__(self, cloudpath: str):
+        self.cloudpath = cloudpath
+        self.type = LayerType.Segmentation
+
+
+def generate_ng_payload(
+    layers: dict[str, Layer], center: Optional[tuple[int, int, int]] = None
+) -> str:
+    """Makes neuroglancer link payloads from mappings between layer names and paths.
+
+    Makes the viewer resolution equal to the "img" layer if it exists in layers.
+    Otherwise, it takes the first it can find.
+    """
+    assert len(layers) > 0, "empty layer paths"
+
+    ngl_layers = OrderedDict()
+    layers = layers.copy()  # for in-place modifications later
+
+    # default to first resolution we can find - update if we find an image
+    ng_resolution = dataset_resolution(list(layers.values())[0])
+
+    # Trying to find an image layer
+    base_img = layers.pop("img", None)
+    if base_img:
+        ngl_layers["img"] = {
+            "source": fullpath(base_img),
+            "type": base_img.type.value,
+        }
+        ng_resolution = dataset_resolution(base_img)
+
+    for key, layer in layers.items():
+        ngl_layers[key] = {
+            "source": fullpath(layer),
+            "type": layer.type.value,
+        }
+
+    navigation = {"pose": {"position": {"voxelSize": ng_resolution}}, "zoomFactor": 4}
+
+    if center is not None:
+        navigation["post"]["position"]["voxelCoordinates"] = center
+
+    payload = OrderedDict(
+        [
+            ("layers", ngl_layers),
+            ("navigation", navigation),
+            ("showSlices", False),
+            ("layout", "xy-3d")
+        ]
+    )
+    return payload
+
+
+def generate_link(
+    layers: dict[str, Layer],
+    host: str = "state-share-dot-neuroglancer-dot-seung-lab.appspot.com",
+) -> str:
+    payload = generate_ng_payload(layers)
+
+    return wrap_payload(payload, host)
+
+
+def wrap_payload(
+    payload: Union[OrderedDict, str],
+    host: str = "state-share-dot-neuroglancer-dot-seung-lab.appspot.com",
```

Comment:
maybe define the default host in a common place, easier to change the host in the future.

Reply:
Yeah I had this in mind too, but I wasn't sure where the best place would be. I think my favorite idea is to put this in the deployment file, but I wasn't sure. What do you think? Probably a good extension anyway.